bmc-acf/acf_manager.cpp

```diff
@@ -0,0 +1,526 @@
+ *  Perform validation of ACF.
+ *  This verifies the signature,
+ *  checks that the serial number stored on ACF matches BMC serial number,
+ *  checks that the expiration on ACF isn't past the current bmc time
+ *
+ *  @param[in] accessControlFile - ACF contents.
+ *
+ *  @param[in] publicKeyFile - Public key contents.
+ *
+ *  @param[out] sDate - Expiration date of ACF file.
+ *
+ *  @return Status of call. Success indicates verification was successful.
+ */
+static CeLogin::CeLoginRc
+    verifyAcfSerialNumberAndExpiration(std::vector<uint8_t>& accessControlFile,
+                                       std::vector<uint8_t>& publicKeyFile,
+                                       std::string& sDate)
+{
+    CeLogin::CELoginSequenceV1* sDecodedAsn = NULL;
+    CeLogin::CeLoginRc sRc = CeLogin::CeLoginRc::Failure;
+    CeLogin::CeLoginJsonData* sJsonData = NULL;
+
+    sRc = decodeAndVerifySignature(
+        accessControlFile.data(), accessControlFile.size(),
+        publicKeyFile.data(), publicKeyFile.size(), sDecodedAsn);
+
+    if (CeLogin::CeLoginRc::Success == sRc)
+    {
+        log<level::INFO>("Signature verified");
+
+        std::string mSerialNumber({});
+        mSerialNumber = readMachineSerialNumberProperty(
+            DBUS_INVENTORY_SYSTEM_OBJECT, DBUS_INVENTORY_ASSET_INTERFACE,
+            DBUS_SERIAL_NUM_PROP);
+
+        // If serial number is empty on machine set as UNSET for check with acf
+        if (mSerialNumber.empty())
+        {
+            mSerialNumber = UNSET_SERIAL_NUM_KEYWORD;
+        }
+        sJsonData = (CeLogin::CeLoginJsonData*)OPENSSL_malloc(
+            sizeof(CeLogin::CeLoginJsonData));
+        if (sJsonData)
+        {
+            memset(sJsonData, 0x00, sizeof(CeLogin::CeLoginJsonData));
+        }
+        else
+        {
+            log<level::ERR>("malloc failed");
+            elog<InternalFailure>();
+        }
+
+        if (CeLogin::CeLoginRc::Success == sRc)
+        {
+            // Verify system serial number is in machine list (and get the
+            // authorization)
+            const uint64_t serialNumberLengthParm = mSerialNumber.length();
+            sRc = CeLogin::decodeJson(
+                (const char*)sDecodedAsn->sourceFileData->data,
+                sDecodedAsn->sourceFileData->length, mSerialNumber.c_str(),
+                serialNumberLengthParm, *sJsonData);
+            if (CeLogin::CeLoginRc::Success == sRc)
+            {
+                log<level::INFO>("Serial Number matches");
+            }
+            else
+            {
+                log<level::ERR>("Serial Number does not match");
+            }
+        }
+    }
+    else if (CeLogin::CeLoginRc::SignatureNotValid == sRc)
+    {
+        log<level::ERR>("Signature is not valid");
+    }
+    else
+    {
+        log<level::ERR>("Error: ", entry("sRc=%d", (int)sRc));
+    }
+    // Check if time is expired
+    if (CeLogin::CeLoginRc::Success == sRc)
+    {
```

Comment:
is this check for sRC for value returend at line 220 are for line no 255? you can choose different variable names in different scopes, this will be very tough to unit test.

Reply:
I nested it after the check for the serial number